README.rst
============
SSL CertInfo
============

.. start-badges

.. list-table::
    :stub-columns: 1

    * - build
      - |travis|
    * - quality
      - |codacy| |codeclimate| |sonar-qg| |sonar-rel|
    * - coverage
      - |coveralls| |codecov| |codeclimate-cov|
    * - dependencies
      - |pyup| |pyup-p3| |requires|
    * - package
      - |version| |pyversions| |downloads|


.. |travis| image:: https://img.shields.io/travis/stdtom/ssl_certinfo/master.svg?logo=travis
   :target: https://travis-ci.com/stdtom/ssl_certinfo
   :alt: Travis Build Status

.. |codacy| image:: https://api.codacy.com/project/badge/Grade/589c03a215ec4ddbb0085b523a857e55
   :target: https://www.codacy.com/manual/stdtom/ssl_certinfo
   :alt: Codacy Grade

.. |codeclimate| image:: https://api.codeclimate.com/v1/badges/1ed86e874b3c68672c5c/maintainability
   :target: https://codeclimate.com/github/stdtom/ssl_certinfo/maintainability
   :alt: Code Climate Maintainability

.. |sonar-qg| image:: https://sonarcloud.io/api/project_badges/measure?project=stdtom_ssl_certinfo&metric=alert_status
   :target: https://sonarcloud.io/dashboard?id=stdtom_ssl_certinfo
   :alt: Sonar Quality Gate Status

.. |sonar-rel| image:: https://sonarcloud.io/api/project_badges/measure?project=stdtom_ssl_certinfo&metric=reliability_rating
   :target: https://sonarcloud.io/dashboard?id=stdtom_ssl_certinfo
   :alt: Sonar Reliability Rating

.. |coveralls| image:: https://coveralls.io/repos/github/stdtom/ssl_certinfo/badge.svg?branch=master
   :target: https://coveralls.io/github/stdtom/ssl_certinfo?branch=master
   :alt: Coveralls Test Coverage

.. |codecov| image:: https://codecov.io/gh/stdtom/ssl_certinfo/branch/master/graph/badge.svg
   :target: https://codecov.io/gh/stdtom/ssl_certinfo
   :alt: CodeCov

.. |codeclimate-cov| image:: https://api.codeclimate.com/v1/badges/1ed86e874b3c68672c5c/test_coverage
   :target: https://codeclimate.com/github/stdtom/ssl_certinfo/test_coverage
   :alt: Code Climate Test Coverage

.. |pyup| image:: https://pyup.io/repos/github/stdtom/ssl_certinfo/shield.svg
   :target: https://pyup.io/repos/github/stdtom/ssl_certinfo/
   :alt: Updates

.. |pyup-p3| image:: https://pyup.io/repos/github/stdtom/ssl_certinfo/python-3-shield.svg
   :target: https://pyup.io/repos/github/stdtom/ssl_certinfo/
   :alt: Python 3

.. |requires| image:: https://requires.io/github/stdtom/ssl_certinfo/requirements.svg?branch=master
   :target: https://requires.io/github/stdtom/ssl_certinfo/requirements/?branch=master
   :alt: Requirements Status

.. |version| image:: https://img.shields.io/pypi/v/ssl-certinfo.svg
   :target: https://pypi.org/project/ssl-certinfo/
   :alt: Version

.. |pyversions| image:: https://img.shields.io/pypi/pyversions/ssl-certinfo.svg?logo=python&logoColor=FBE072
    :target: https://pypi.org/project/ssl-certinfo/
    :alt: Python versions supported

.. |downloads| image:: https://pepy.tech/badge/ssl-certinfo
    :target: https://pepy.tech/project/ssl-certinfo
    :alt: PyPI downloads

.. end-badges


SSL CertInfo collects information about SSL certificates from a set of hosts.


Features
--------

* Hosts to be scanned can be specified as a list of

  * hostnames (fully qualified domain names), e.g. ``github.com``,
  * ip addresses, e.g. ``1.1.1.1``,
  * ip networks in CIDR format, e.g. ``10.0.0.0/24``,
  * ip ranges, e.g. ``10.0.0.1-10.0.0.10``,
  * or any combination of the previous.

* Results will be presented in various output formats: ``--table``, ``--json``, ``--yaml``, ``--csv``, ``--raw``.


Installation
------------
You can download and install the latest version of this software from the Python package index (PyPI) as follows::

  $ pip install ssl_certinfo


Usage
-----

When you install ssl_certinfo, a command-line script called ``ssl_certinfo`` is
placed on your path. You can invoke ssl_certinfo directly via this script from the command line::

  $ ssl_certinfo [...]


You can also invoke it through the Python interpreter from the command line::

  $ python -m ssl_certinfo [...]


Help is available with the ``--help`` or ``-h`` switch::

  $ ssl_certinfo -h
  usage: ssl_certinfo [-h] [-V] [-v | -q] [-p PORT] [-t TIMEOUT] [-T | -j | -y | -c | -r] [host [host ...]]

  Collect information about SSL certificates from a set of hosts

  positional arguments:
  host                  Connect to HOST

  optional arguments:
  -h, --help            show this help message and exit
  -V, --version         display version information and exit
  -v, --verbose         verbose output (repeat for increased verbosity)
  -q, --quiet           quiet output (show errors only)
  -p PORT, --port PORT  TCP port to connnect to [0-65535]
  -t TIMEOUT, --timeout TIMEOUT
                        Maximum time allowed for connection
  -T, --table           Print results in table format
  -j, --json            Print results in JSON format
  -y, --yaml            Print results in YAML format
  -c, --csv             Print results in CSV format
  -r, --raw             Print results in raw format




Credits
-------

This package was created with Cookiecutter_ and the `stdtom/cookiecutter-pypackage-pipenv`_ project template, based on `audreyr/cookiecutter-pypackage`_.

.. _Cookiecutter: https://github.com/cookiecutter/cookiecutter
.. _`stdtom/cookiecutter-pypackage-pipenv`: https://github.com/stdtom/cookiecutter-pypackage-pipenv
.. _`audreyr/cookiecutter-pypackage`: https://github.com/audreyr/cookiecutter-pypackage


HISTORY.rst
=======
History
=======


see `CHANGELOG.md <./CHANGELOG.md>`_


tests/__init__.py
"""Unit test package for ssl_certinfo."""


tests/test_cli.py
#!/usr/bin/env python

"""Unit test for `ssl_certinfo.cli` module.

Use tox or py.test to run the test suite.
"""

import subprocess
from argparse import ArgumentTypeError

import pytest

from ssl_certinfo import __author__, __email__, __version__, cli
from ssl_certinfo.ssl_certinfo import OutputFormat


@pytest.fixture
def parser():
    return cli.create_parser()


@pytest.mark.parametrize("test_input", [1, 2, 65535, "2"])
def test_valid_positive(test_input):
    assert cli.check_positive(test_input)


@pytest.mark.parametrize("test_input", [-1, -10, "x"])
def test_invalid_positive(test_input):
    with pytest.raises(ArgumentTypeError):
        assert cli.check_positive(test_input)


@pytest.mark.parametrize(
    "args,expected,comment",
    [
        (["github.com"], 5, "default timeout 5 sec"),
        (["github.com", "-t", "2"], 2, "timeout 2 sec"),
    ],
)
def test_cli_valid_timeout(parser, args, expected, comment):
    """Sample pytest test function with the pytest fixture as an argument."""
    args = parser.parse_args(args)
    assert args.timeout == expected


@pytest.mark.parametrize(
    "args,expected,comment",
    [(["github.com", "-t", "-1"], -1, "invalid timeout -1 sec")],
)
def test_cli_invalid_timeout(parser, args, expected, comment):
    """Sample pytest test function with the pytest fixture as an argument."""
    with pytest.raises(SystemExit):
        args = parser.parse_args(args)


@pytest.mark.parametrize("test_input", [1, 2, 65535, "2"])
def test_valid_port(test_input):
    assert cli.check_valid_port(test_input)


@pytest.mark.parametrize("test_input", [-1, 65536, "x"])
def test_invalid_port(test_input):
    with pytest.raises(ArgumentTypeError):
        assert cli.check_valid_port(test_input)


@pytest.mark.parametrize(
    "args,expected,comment",
    [
        (["github.com"], 443, "default port 443"),
        (["github.com", "-p", "8443"], 8443, "port 8443"),
    ],
)
def test_cli_valid_port(parser, args, expected, comment):
    """Sample pytest test function with the pytest fixture as an argument."""
    args = parser.parse_args(args)
    assert args.port == expected


@pytest.mark.parametrize(
    "args,comment",
    [
        (["github.com", "-p", "-1"], "invalid port -1"),
        (["github.com", "-p", "65536"], "invalid port 65536"),
    ],
)
def test_cli_invalid_port(parser, args, comment):
    """Sample pytest test function with the pytest fixture as an argument."""
    with pytest.raises(SystemExit):
        args = parser.parse_args(args)


@pytest.mark.parametrize(
    "inlist,expected,comment",
    [
        (
            "github.com 1.1.1.1".split(" "),
            ["github.com", "1.1.1.1"],
            "nothing to expand",
        ),
        (
            "github.com 192.168.0.0/30 1.1.1.1".split(" "),
            [
                "github.com",
                "192.168.0.0",
                "192.168.0.1",
                "192.168.0.2",
                "192.168.0.3",
                "1.1.1.1",
            ],
            "expand /30 network",
        ),
        (
            ["github.com", "192.168.0.253 - 192.168.1.2", "1.1.1.1"],
            [
                "github.com",
                "192.168.0.253",
                "192.168.0.254",
                "192.168.0.255",
                "192.168.1.0",
                "192.168.1.1",
                "192.168.1.2",
                "1.1.1.1",
            ],
            "expand ip range across subnet boundaries",
        ),
    ],
)
def test_expand_hosts(inlist, expected, comment):
    out = cli.expand_hosts(inlist)
    assert out == expected


@pytest.mark.parametrize(
    "inlist,expected,comment",
    [
        pytest.param(
            "23.1.1.1/8".split(" "),
            256 ** 3,
            "expand class A network",
            marks=pytest.mark.skip,
        ),
        ("130.80.0.0/16".split(" "), 65536, "expand class B network",),
        ("192.168.0.0/24".split(" "), 256, "expand class C network",),
        ("192.168.0.0/30".split(" "), 4, "expand /30 network",),
        ("192.168.0.0-192.168.1.255".split(" "), 512, "range of 2 class C networks",),
    ],
)
def test_expand_hosts_large_networks(inlist, expected, comment):
    out = cli.expand_hosts(inlist)
    assert len(out) == expected


@pytest.mark.parametrize(
    "args,expected,comment",
    [
        ("github.com".split(" "), ["github.com"], "valid host github.com"),
        ("1.1.1.1".split(" "), ["1.1.1.1"], "valid ip address 1.1.1.1"),
        (
            "github.com 1.1.1.1".split(" "),
            ["github.com", "1.1.1.1"],
            "two targets: valid hostname and ip address",
        ),
        ("192.0.2.0/24".split(" "), ["192.0.2.0/24"], "valid ip network 192.0.2.0/24"),
        (
            "10.0.0.1-10.0.0.5".split(" "),
            ["10.0.0.1-10.0.0.5"],
            "valid ip range 10.0.0.1 - .5",
        ),
    ],
)
def test_cli_valid_host_or_ip(parser, args, expected, comment):
    """Sample pytest test function with the pytest fixture as an argument."""
    args = parser.parse_args(args)
    assert args.host == expected


@pytest.mark.parametrize(
    "args,comment",
    [
        (["github.com-"], "invalid hostwith trailing dash"),
        (["1.1.1.256"], "invalid ip address 1.1.1.256"),
    ],
)
def test_cli_indvalid_host_or_ip(parser, args, comment):
    """Sample pytest test function with the pytest fixture as an argument."""
    with pytest.raises(SystemExit):
        args = parser.parse_args(args)


@pytest.mark.parametrize(
    "args,expected,comment",
    [
        ("github.com".split(), OutputFormat.TABLE, "default"),
        ("github.com --table".split(), OutputFormat.TABLE, "TABLE"),
        ("github.com -T".split(), OutputFormat.TABLE, "TABLE"),
        ("github.com --json".split(), OutputFormat.JSON, "JSON"),
        ("github.com -j".split(), OutputFormat.JSON, "JSON"),
        ("github.com --yaml".split(), OutputFormat.YAML, "YAML"),
        ("github.com -y".split(), OutputFormat.YAML, "YAML"),
        ("github.com --csv".split(), OutputFormat.CSV, "CSV"),
        ("github.com -c".split(), OutputFormat.CSV, "CSV"),
        ("github.com --raw".split(), OutputFormat.RAW, "RAW"),
        ("github.com -r".split(), OutputFormat.RAW, "RAW"),
    ],
)
def test_cli_outform(parser, args, expected, comment):
    """Sample pytest test function with the pytest fixture as an argument."""
    args = parser.parse_args(args)
    assert args.outform == expected


def capture(command):
    proc = subprocess.Popen(command, stdout=subprocess.PIPE, stderr=subprocess.PIPE,)
    out, err = proc.communicate()
    return out, err, proc.returncode


def test_cli_main_version():
    command = "python -m ssl_certinfo -V".split(" ")
    out, err, exitcode = capture(command)
    assert exitcode == 0
    assert out.decode().find(__version__) >= 0
    assert out.decode().find(__author__) >= 0
    assert out.decode().find(__email__) >= 0


def test_cli_main_single_target():
    command = "python -m ssl_certinfo github.com".split(" ")
    out, err, exitcode = capture(command)
    assert exitcode == 0
    assert out.decode().find("github") >= 0
    assert (err == b"") or (err.decode().find("100%") >= 0)


def test_cli_main_two_targets():
    command = "python -m ssl_certinfo github.com wikipedia.org".split(" ")
    out, err, exitcode = capture(command)
    assert exitcode == 0
    assert out.decode().find("github") >= 0
    assert out.decode().find("wikipedia") >= 0
    assert (err == b"") or (err.decode().find("100%") >= 0)


tests/test_validation.py
#!/usr/bin/env python

"""Unit test for `validation` package.

Use tox or py.test to run the test suite.
"""

import pytest

from ssl_certinfo import validation


@pytest.mark.parametrize("test_input", ["1.0.0.1", "255.255.255.255"])
def test_valid_ip_address(test_input):
    assert validation.is_valid_ip_address(test_input)


@pytest.mark.parametrize("test_input", ["10.0.0.-1", "10.0.0.1/33"])
def test_invalid_ip_address(test_input):
    assert not validation.is_valid_ip_address(test_input)


@pytest.mark.parametrize(
    "test_input",
    [
        "1.0.0.1/8",
        "1.0.0.1/255.0.0.0",
        "4.4.0.0/16",
        "4.4.0.0/255.255.0.0",
        "192.0.2.0/24",
        "192.0.2.0/255.255.255.0",
    ],
)
def test_valid_ip_network(test_input):
    assert validation.is_valid_ip_network(test_input)


@pytest.mark.parametrize("test_input", ["10.0.0.1", "10.0.0.1/33"])
def test_invalid_ip_network(test_input):
    assert not validation.is_valid_ip_network(test_input)


@pytest.mark.parametrize(
    "test_input", ["10.0.0.1 - 10.0.0.5", "192.168.0.1-192.168.1.255"]
)
def test_valid_ip_range(test_input):
    assert validation.is_valid_ip_range(test_input)


@pytest.mark.parametrize(
    "test_input",
    [
        "10.0.0.1",  # single ip address, no range
        "10.0.0.0/24",  # ip network in cidr
        "10.0.0.5 - 10.0.0.1",  # start ip greater than end ip
        "192.168.0.1-192.168.0.1",  # range of only one ip address
        "10.0.0.1/32 - 10.0.0.5",  # start ip is cidr
        "a.b.c - 1.1.1.1",  # start is not a valid ip
        "abcde",  # characters
    ],
)
def test_invalid_ip_range(test_input):
    assert not validation.is_valid_ip_range(test_input)


@pytest.mark.parametrize(
    "test_input",
    [
        "github.com",
        "www." + ("abcdefghi" * 7) + ".com",  # second level domain with 63 chars
        ("abcdefghi" * 7 + ".") * 3,  # 3 labels with 63 chars plus trailing '.'
        ("a" * 49 + ".") * 5 + "com",  # total length 253 chars
        "abc.com",
    ],
)
def test_valid_hostname(test_input):
    assert validation.is_valid_hostname(test_input)


@pytest.mark.parametrize(
    "test_input",
    [
        "www." + ("abcdefghi" * 7) + "x.com",  # second level domain with 64 chars
        ("a" * 49 + ".") * 5 + "xcom",  # total length 254 chars
        "-abc.com",
        "abc-.com",
        "-abc-.com",
        "abc.-com",
        "abc.com-",
        "abc.-com-",
        "10.0.0.-1",
        "10.0.0.1/33",
        "10.0.0.1",
        "255.255.255.255",
    ],
)
def test_invalid_hostname(test_input):
    assert not validation.is_valid_hostname(test_input)


ssl_certinfo/validation.py
"""Validate ip address and hostname."""
import ipaddress
import re


def is_valid_ip_address(value):
    """Validate if parameter is a valid ip address."""
    try:
        ipaddress.ip_address(value)
        return True
    except ValueError:
        return False


def is_valid_ip_network(value):
    """Validate if parameter is a valid ip network."""
    try:
        net = ipaddress.ip_network(value, False)
    except ValueError:
        return False
    else:
        if net.prefixlen < 32:
            return True
        else:
            return False


def is_valid_ip_range(value):
    """Validate if parameter is a valid ip address range."""
    separator = re.compile(r" *- *")
    try:
        (start, end) = separator.split(value)
        start_addr = ipaddress.ip_address(start)
        end_addr = ipaddress.ip_address(end)
    except ValueError:
        return False
    else:
        if start_addr < end_addr:
            return True
        else:
            return False


def is_valid_hostname(hostname):
    """Validate if parameter is a valid fqdn as defined in RFC-1035."""
    if hostname[-1] == ".":
        # strip exactly one dot from the right, if present
        hostname = hostname[:-1]
    if len(hostname) > 253:
        return False

    labels = hostname.split(".")

    # the TLD must be not all-numeric
    if re.match(r"[0-9]+$", labels[-1]):
        return False

    allowed = re.compile("^[a-z0-9]([a-z0-9-]{0,61}[a-z0-9])?$", re.IGNORECASE)
    return all(allowed.match(label) for label in labels)


ssl_certinfo/ssl_certinfo.py
"""Main module."""
import enum
import json
import logging
import time
from datetime import datetime
from socket import socket

import pandas as pd
import yaml
from cryptography import x509
from cryptography.x509.oid import NameOID
from OpenSSL import SSL
from OpenSSL.SSL import WantReadError, WantWriteError
from tabulate import tabulate
from tqdm import tqdm


class OutputFormat(enum.Enum):
    TABLE = 1
    JSON = 2
    YAML = 3
    CSV = 4
    RAW = 5


def get_cert_info(cert):
    """Get all information about SSL certificate ."""
    certinfo = {}

    certinfo["CN"] = cert.subject.get_attributes_for_oid(NameOID.COMMON_NAME)[0].value

    ext = cert.extensions.get_extension_for_class(x509.SubjectAlternativeName)
    san_list = ext.value.get_values_for_type(x509.DNSName)
    certinfo["SAN"] = ";".join(san_list)

    certinfo["valid_from"] = cert.not_valid_before.isoformat()
    certinfo["valid_to"] = cert.not_valid_after.isoformat()

    delta = cert.not_valid_after - datetime.now()
    certinfo["expire_in_days"] = delta.days

    return certinfo


def ssl_handshake_helper(sock_ssl):
    timeout = sock_ssl.gettimeout()
    if timeout is not None:
        start = time.time()
    while True:
        try:
            return sock_ssl.do_handshake()
        except (WantReadError, WantWriteError):
            if start + timeout <= time.time():
                raise TimeoutError


def get_certificate(hostname, port, timeout=5):
    sock = socket()
    sock.settimeout(timeout)
    sock.connect((hostname, port))

    context = SSL.Context(SSL.SSLv23_METHOD)

    context.check_hostname = False
    context.verify_mode = SSL.VERIFY_NONE

    sock_ssl = SSL.Connection(context, sock)
    sock_ssl.set_connect_state()
    sock_ssl.set_tlsext_host_name(hostname.encode())
    ssl_handshake_helper(sock_ssl)
    cert = sock_ssl.get_peer_certificate()

    sock_ssl.close()
    sock.close()

    return cert.to_cryptography()


def result_to_dataframe(result_dict):
    column_names = [
        "CN",
        "SAN",
        "valid_from",
        "valid_to",
        "expire_in_days",
        "peername",
        "peerport",
    ]
    df = pd.DataFrame(result_dict).T.rename_axis("peer", axis=1)
    df = df.reindex(columns=column_names)

    return df


def process_hosts(hosts, default_port, timeout=5, outform=OutputFormat.TABLE):
    results = {}

    progbar = tqdm(hosts)

    for host in progbar:
        progbar.set_description(f"Checking {host}...")
        try:
            logging.info("Trying to fetch certificate for " + host)
            cert = get_certificate(host, default_port, timeout)
        except (OSError, SSL.Error):
            logging.info("Could not fetch certificate for " + host)
        else:
            certinfo = get_cert_info(cert)
            certinfo["peername"] = host
            certinfo["peerport"] = default_port

            results[host] = certinfo
    print(format_results(results, outform))


def format_results(results, outform):
    if results == {}:
        return ""

    elif outform == OutputFormat.TABLE:
        df = result_to_dataframe(results)
        df.index.name = "peer"
        return tabulate(df, headers="keys", tablefmt="pretty")

    elif outform == OutputFormat.JSON:
        return json.dumps(results, indent=4)

    elif outform == OutputFormat.YAML:
        return yaml.dump(results)

    elif outform == OutputFormat.CSV:
        df = result_to_dataframe(results)
        df.index.name = "peer"
        return df.to_csv()

    elif outform == OutputFormat.RAW:
        df = result_to_dataframe(results)
        return df.to_string()


ssl_certinfo/cli.py
"""Console script for ssl_certinfo."""
import argparse
import ipaddress
import logging
import re
import sys

from ssl_certinfo import __author__, __email__, __version__, ssl_certinfo, validation
from ssl_certinfo.ssl_certinfo import OutputFormat

VERSION = rf"""
ssl_certinfo {__version__}
Copyright (C) 2020 {__author__} ({__email__})
License Apache-2.0: <http://www.apache.org/licenses/LICENSE-2.0>.
"""


def check_hostname_or_ip_address(value):
    """Validate argparse type hostname/ip address."""
    if (
        not validation.is_valid_hostname(value)
        and not validation.is_valid_ip_address(value)
        and not validation.is_valid_ip_network(value)
        and not validation.is_valid_ip_range(value)
    ):
        raise argparse.ArgumentTypeError(
            "%s is not a valid hostname or ip address" % value
        )
    return value


def check_positive(value):
    """Validate argparse type positive integer."""
    try:
        ivalue = int(value)
    except ValueError:
        raise argparse.ArgumentTypeError("%s is not an int value" % value)

    if ivalue <= 0:
        raise argparse.ArgumentTypeError("%s is not a positive int value" % value)
    return ivalue


def check_valid_port(value):
    """Validate argparse type TCP port number."""
    try:
        ivalue = int(value)
    except ValueError:
        raise argparse.ArgumentTypeError("%s is an invalid port number" % value)

    if ivalue <= 0 or ivalue > 65535:
        raise argparse.ArgumentTypeError("%s is an invalid port number" % value)
    return ivalue


def expand_hosts(hostlist):
    result = []

    for elem in hostlist:
        if validation.is_valid_hostname(elem) or validation.is_valid_ip_address(elem):
            result.append(elem)
        elif validation.is_valid_ip_range(elem):
            separator = re.compile(r" *- *")
            (start, end) = separator.split(elem)
            try:
                start_addr = ipaddress.ip_address(start)
                end_addr = ipaddress.ip_address(end)
            except ValueError:
                pass
            else:
                logging.debug("Expanding ip address range " + elem)
                current_ipaddr = start_addr
                while current_ipaddr <= end_addr:
                    result.append(str(current_ipaddr))
                    current_ipaddr = current_ipaddr + 1
        else:
            try:
                net = ipaddress.ip_network(elem, False)
            except ValueError:
                pass
            else:
                logging.debug("Expanding ip network " + elem)
                for ipaddr in net:
                    result.append(str(ipaddr))

    return result


def create_parser():
    """Create ArgParser."""
    parser = argparse.ArgumentParser(
        description="Collect information about SSL certificates from a set of hosts"
    )

    parser.add_argument(
        "-V",
        "--version",
        action="store_true",
        dest="displayVersion",
        help="display version information and exit",
    )

    verb_group = parser.add_mutually_exclusive_group()
    verb_group.add_argument(
        "-v",
        "--verbose",
        action="count",
        dest="verbosity",
        default=0,
        help="verbose output (repeat for increased verbosity)",
    )
    verb_group.add_argument(
        "-q",
        "--quiet",
        action="store_const",
        const=-1,
        default=0,
        dest="verbosity",
        help="quiet output (show errors only)",
    )

    parser.add_argument(
        "host", nargs="*", type=check_hostname_or_ip_address, help="Connect to HOST",
    )

    parser.add_argument(
        "-p",
        "--port",
        default=443,
        type=check_valid_port,
        help="TCP port to connnect to [0-65535]",
    )

    parser.add_argument(
        "-t",
        "--timeout",
        default=5,
        type=check_positive,
        help="Maximum time allowed for connection",
    )

    output_format = parser.add_mutually_exclusive_group()
    output_format.add_argument(
        "-T",
        "--table",
        action="store_const",
        const=OutputFormat.TABLE,
        default=OutputFormat.TABLE,
        dest="outform",
        help="Print results in table format",
    )
    output_format.add_argument(
        "-j",
        "--json",
        action="store_const",
        const=OutputFormat.JSON,
        default=OutputFormat.TABLE,
        dest="outform",
        help="Print results in JSON format",
    )
    output_format.add_argument(
        "-y",
        "--yaml",
        action="store_const",
        const=OutputFormat.YAML,
        default=OutputFormat.TABLE,
        dest="outform",
        help="Print results in YAML format",
    )
    output_format.add_argument(
        "-c",
        "--csv",
        action="store_const",
        const=OutputFormat.CSV,
        default=OutputFormat.TABLE,
        dest="outform",
        help="Print results in CSV format",
    )
    output_format.add_argument(
        "-r",
        "--raw",
        action="store_const",
        const=OutputFormat.RAW,
        default=OutputFormat.TABLE,
        dest="outform",
        help="Print results in raw format",
    )

    return parser


def setup_logging(verbosity):
    base_loglevel = 30
    verbosity = min(verbosity, 2)
    loglevel = base_loglevel - (verbosity * 10)
    logging.basicConfig(level=loglevel, format="%(levelname)s\t%(message)s")


def main():
    """Console script for ssl_certinfo."""
    args = create_parser().parse_args()
    if args.displayVersion:
        print(VERSION)
        return 0

    setup_logging(args.verbosity)

    logging.info("Arguments: " + str(args))

    ssl_certinfo.process_hosts(
        expand_hosts(args.host), args.port, args.timeout, args.outform
    )
    return 0


if __name__ == "__main__":
    sys.exit(main())  # pragma: no cover
